{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98987003",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "28d3e693",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/fs/scratch/PAS0439/Ming/virome_ecology_core_prkaryotes/results/08_inStrain_rug4941/prophage_carrying_genome_annotate/dram_anno/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "309083f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "prophage_co_exist_microdiversity = pd.read_csv(\"/fs/ess/PAS0439/MING/virome_ecology/results/R_project/prophage_co_exist_microdiversity_rug4941_tmp.csv\")\n",
    "prophage_microdiversity = pd.read_csv(\"/fs/ess/PAS0439/MING/virome_ecology/results/R_project/prophage_microdiversity_rug4941.csv_tmp\")\n",
    "instrain_genome_taxa = pd.read_csv(\"../../genome_info.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "78f2136d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3c19528c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_under_selection = set(prophage_microdiversity.query('prophage_gene == \"Yes\"').query('SNV_count >= 10').query('pNpS_variants > 1').gene)\n",
    "n_gene_under_selection = len(gene_under_selection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "72f82e94",
   "metadata": {},
   "outputs": [],
   "source": [
    "records = SeqIO.parse(\"../../repre_genomes_instrain.genes.faa\", \"fasta\")\n",
    "with open(\"/fs/scratch/PAS0439/Ming/virome_ecology_core_prkaryotes/results/08_inStrain_rug4941/prophage_carrying_genome_annotate/gene_under_selection.faa\", \"w\") as outfile:\n",
    "    for record in records:\n",
    "        if record.id in gene_under_selection:\n",
    "            SeqIO.write(record, outfile, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d6209f8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_gene_qualified = len(set(prophage_microdiversity.query('prophage_gene == \"Yes\"').query('SNV_count >= 10').gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a2977cd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Out of 1317 genes, 129 are under postive selection'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f\"Out of {n_gene_qualified} genes, {n_gene_under_selection} are under postive selection\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b2c7e6",
   "metadata": {},
   "source": [
    "*For species which both prophage-carrying and non-carrying population co-exist, whether prophage gene help specification*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f4bc6f49",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prophage_co_exist_microdiversity.query('prophage_gene == \"Yes\"').query('SNV_count > 10').query('pNpS_variants > 1').groupby(\"gene\").id.nunique(\"id\").reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03e0f154",
   "metadata": {},
   "source": [
    "*For each prophage-carrying population among the co-exist species, how many samples they are presented*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "041befb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_902755265.1_Rumen_uncultured_genome_RUG10044_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902755475.1_Rumen_uncultured_genome_RUG10070_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902755575.1_Rumen_uncultured_genome_RUG10072_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902755765.1_Rumen_uncultured_genome_RUG10087_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902756485.1_Rumen_uncultured_genome_RUG10164_genomic</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902757265.1_Rumen_uncultured_genome_RUG10241_genomic</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902758485.1_Rumen_uncultured_genome_RUG10369_genomic</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902759005.1_Rumen_uncultured_genome_RUG10416_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902760985.1_Rumen_uncultured_genome_RUG10617_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902762165.1_Rumen_uncultured_genome_RUG10730_genomic</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902764155.1_Rumen_uncultured_genome_RUG10935_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902764265.1_Rumen_uncultured_genome_RUG10939_genomic</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902764425.1_Rumen_uncultured_genome_RUG10959_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902764875.1_Rumen_uncultured_genome_RUG10998_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902764905.1_Rumen_uncultured_genome_RUG11010_genomic</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902765815.1_Rumen_uncultured_genome_RUG11098_genomic</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902769495.1_Rumen_uncultured_genome_RUG11472_genomic</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902770115.1_Rumen_uncultured_genome_RUG11533_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902770465.1_Rumen_uncultured_genome_RUG11578_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902771395.1_Rumen_uncultured_genome_RUG11664_genomic</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902771415.1_Rumen_uncultured_genome_RUG11667_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902773665.1_Rumen_uncultured_genome_RUG11894_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902774595.1_Rumen_uncultured_genome_RUG11984_genomic</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902774805.1_Rumen_uncultured_genome_RUG12010_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902775065.1_Rumen_uncultured_genome_RUG12030_genomic</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902775115.1_Rumen_uncultured_genome_RUG12039_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902775535.1_Rumen_uncultured_genome_RUG12076_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902775975.1_Rumen_uncultured_genome_RUG12118_genomic</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902777305.1_Rumen_uncultured_genome_RUG12257_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic</th>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902777725.1_Rumen_uncultured_genome_RUG12297_genomic</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902778235.1_Rumen_uncultured_genome_RUG12351_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902779635.1_Rumen_uncultured_genome_RUG12490_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902779725.1_Rumen_uncultured_genome_RUG12498_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902780775.1_Rumen_uncultured_genome_RUG12601_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902784365.1_Rumen_uncultured_genome_RUG12957_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902785415.1_Rumen_uncultured_genome_RUG13069_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902785555.1_Rumen_uncultured_genome_RUG13073_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902785565.1_Rumen_uncultured_genome_RUG13078_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902788035.1_Rumen_uncultured_genome_RUG13328_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902788215.1_Rumen_uncultured_genome_RUG13348_genomic</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902789115.1_Rumen_uncultured_genome_RUG13439_genomic</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902790165.1_Rumen_uncultured_genome_RUG13548_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902790195.1_Rumen_uncultured_genome_RUG13545_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902790765.1_Rumen_uncultured_genome_RUG13606_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902793305.1_Rumen_uncultured_genome_RUG13854_genomic</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902793665.1_Rumen_uncultured_genome_RUG13894_genomic</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902793885.1_Rumen_uncultured_genome_RUG13918_genomic</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902794805.1_Rumen_uncultured_genome_RUG14004_genomic</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902794915.1_Rumen_uncultured_genome_RUG14016_genomic</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902798575.1_Rumen_uncultured_genome_RUG14384_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902798755.1_Rumen_uncultured_genome_RUG14402_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902799065.1_Rumen_uncultured_genome_RUG14430_genomic</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902799085.1_Rumen_uncultured_genome_RUG14432_genomic</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902799195.1_Rumen_uncultured_genome_RUG14443_genomic</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902800535.1_Rumen_uncultured_genome_RUG14574_genomic</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902801305.1_Rumen_uncultured_genome_RUG14652_genomic</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902801315.1_Rumen_uncultured_genome_RUG14650_genomic</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_902802865.1_Rumen_uncultured_genome_RUG14812_genomic</th>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    id\n",
       "genome                                                \n",
       "GCA_902755265.1_Rumen_uncultured_genome_RUG1004...   1\n",
       "GCA_902755475.1_Rumen_uncultured_genome_RUG1007...   1\n",
       "GCA_902755575.1_Rumen_uncultured_genome_RUG1007...   1\n",
       "GCA_902755765.1_Rumen_uncultured_genome_RUG1008...   1\n",
       "GCA_902756485.1_Rumen_uncultured_genome_RUG1016...   3\n",
       "GCA_902757265.1_Rumen_uncultured_genome_RUG1024...  15\n",
       "GCA_902758485.1_Rumen_uncultured_genome_RUG1036...   4\n",
       "GCA_902759005.1_Rumen_uncultured_genome_RUG1041...   1\n",
       "GCA_902760985.1_Rumen_uncultured_genome_RUG1061...   1\n",
       "GCA_902762165.1_Rumen_uncultured_genome_RUG1073...   9\n",
       "GCA_902764155.1_Rumen_uncultured_genome_RUG1093...   2\n",
       "GCA_902764265.1_Rumen_uncultured_genome_RUG1093...   3\n",
       "GCA_902764425.1_Rumen_uncultured_genome_RUG1095...   1\n",
       "GCA_902764875.1_Rumen_uncultured_genome_RUG1099...   2\n",
       "GCA_902764905.1_Rumen_uncultured_genome_RUG1101...  13\n",
       "GCA_902765815.1_Rumen_uncultured_genome_RUG1109...  14\n",
       "GCA_902769495.1_Rumen_uncultured_genome_RUG1147...  17\n",
       "GCA_902770115.1_Rumen_uncultured_genome_RUG1153...   1\n",
       "GCA_902770465.1_Rumen_uncultured_genome_RUG1157...   1\n",
       "GCA_902771395.1_Rumen_uncultured_genome_RUG1166...   6\n",
       "GCA_902771415.1_Rumen_uncultured_genome_RUG1166...   1\n",
       "GCA_902773665.1_Rumen_uncultured_genome_RUG1189...   1\n",
       "GCA_902774595.1_Rumen_uncultured_genome_RUG1198...   6\n",
       "GCA_902774805.1_Rumen_uncultured_genome_RUG1201...   1\n",
       "GCA_902775065.1_Rumen_uncultured_genome_RUG1203...  33\n",
       "GCA_902775115.1_Rumen_uncultured_genome_RUG1203...   1\n",
       "GCA_902775535.1_Rumen_uncultured_genome_RUG1207...   1\n",
       "GCA_902775975.1_Rumen_uncultured_genome_RUG1211...   6\n",
       "GCA_902777305.1_Rumen_uncultured_genome_RUG1225...   1\n",
       "GCA_902777695.1_Rumen_uncultured_genome_RUG1229...  46\n",
       "GCA_902777725.1_Rumen_uncultured_genome_RUG1229...  33\n",
       "GCA_902778235.1_Rumen_uncultured_genome_RUG1235...   1\n",
       "GCA_902779635.1_Rumen_uncultured_genome_RUG1249...   2\n",
       "GCA_902779725.1_Rumen_uncultured_genome_RUG1249...   2\n",
       "GCA_902780775.1_Rumen_uncultured_genome_RUG1260...   2\n",
       "GCA_902784365.1_Rumen_uncultured_genome_RUG1295...   2\n",
       "GCA_902785415.1_Rumen_uncultured_genome_RUG1306...   1\n",
       "GCA_902785555.1_Rumen_uncultured_genome_RUG1307...   1\n",
       "GCA_902785565.1_Rumen_uncultured_genome_RUG1307...   2\n",
       "GCA_902788035.1_Rumen_uncultured_genome_RUG1332...   1\n",
       "GCA_902788215.1_Rumen_uncultured_genome_RUG1334...  11\n",
       "GCA_902789115.1_Rumen_uncultured_genome_RUG1343...  12\n",
       "GCA_902790165.1_Rumen_uncultured_genome_RUG1354...   2\n",
       "GCA_902790195.1_Rumen_uncultured_genome_RUG1354...   1\n",
       "GCA_902790765.1_Rumen_uncultured_genome_RUG1360...   1\n",
       "GCA_902793305.1_Rumen_uncultured_genome_RUG1385...  24\n",
       "GCA_902793665.1_Rumen_uncultured_genome_RUG1389...   8\n",
       "GCA_902793885.1_Rumen_uncultured_genome_RUG1391...  10\n",
       "GCA_902794805.1_Rumen_uncultured_genome_RUG1400...   4\n",
       "GCA_902794915.1_Rumen_uncultured_genome_RUG1401...   3\n",
       "GCA_902798575.1_Rumen_uncultured_genome_RUG1438...   1\n",
       "GCA_902798755.1_Rumen_uncultured_genome_RUG1440...   1\n",
       "GCA_902799065.1_Rumen_uncultured_genome_RUG1443...   4\n",
       "GCA_902799085.1_Rumen_uncultured_genome_RUG1443...   6\n",
       "GCA_902799195.1_Rumen_uncultured_genome_RUG1444...  29\n",
       "GCA_902800535.1_Rumen_uncultured_genome_RUG1457...   1\n",
       "GCA_902801305.1_Rumen_uncultured_genome_RUG1465...  30\n",
       "GCA_902801315.1_Rumen_uncultured_genome_RUG1465...   2\n",
       "GCA_902802865.1_Rumen_uncultured_genome_RUG1481...  20"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prophage_co_exist_microdiversity.query('prophage_gene == \"Yes\"')[[\"genome\", \"id\"]].drop_duplicates().groupby(\"genome\").nunique(\"id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "110962ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>user_genome</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genera</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2227</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Bacteroidaceae</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>Prevotella sp900317685</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            user_genome    Domain  \\\n",
       "2227  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  Bacteria   \n",
       "\n",
       "            Phylum        Class          Order          Family      Genera  \\\n",
       "2227  Bacteroidota  Bacteroidia  Bacteroidales  Bacteroidaceae  Prevotella   \n",
       "\n",
       "                     Species  \n",
       "2227  Prevotella sp900317685  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instrain_genome_taxa.query('user_genome == \"GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5edecda3",
   "metadata": {},
   "source": [
    "*Prophage-carrying genome (MAG), GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic.fasta  \n",
    "co-exist with prophage-non-carrying MAGs (of the same species: Prevotella sp900317685) in 46 samples*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a55074e",
   "metadata": {},
   "source": [
    "*For GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic.fasta, any gene under selection?*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "781c5420",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>0.117536</td>\n",
       "      <td>0.148610</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027985</td>\n",
       "      <td>0.279848</td>\n",
       "      <td>0.279848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_12</th>\n",
       "      <td>19.0</td>\n",
       "      <td>0.083405</td>\n",
       "      <td>0.047692</td>\n",
       "      <td>0.031691</td>\n",
       "      <td>0.060287</td>\n",
       "      <td>0.072677</td>\n",
       "      <td>0.081762</td>\n",
       "      <td>0.233605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_13</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.019184</td>\n",
       "      <td>0.016632</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013999</td>\n",
       "      <td>0.027999</td>\n",
       "      <td>0.028777</td>\n",
       "      <td>0.029554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_15</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.187434</td>\n",
       "      <td>0.063689</td>\n",
       "      <td>0.135379</td>\n",
       "      <td>0.151925</td>\n",
       "      <td>0.168472</td>\n",
       "      <td>0.213461</td>\n",
       "      <td>0.258451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_17</th>\n",
       "      <td>9.0</td>\n",
       "      <td>0.178138</td>\n",
       "      <td>0.044825</td>\n",
       "      <td>0.099832</td>\n",
       "      <td>0.168467</td>\n",
       "      <td>0.192534</td>\n",
       "      <td>0.199665</td>\n",
       "      <td>0.232942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_2</th>\n",
       "      <td>13.0</td>\n",
       "      <td>1.853915</td>\n",
       "      <td>0.896812</td>\n",
       "      <td>0.477837</td>\n",
       "      <td>1.314051</td>\n",
       "      <td>1.791888</td>\n",
       "      <td>2.150266</td>\n",
       "      <td>3.942154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_3</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.030396</td>\n",
       "      <td>0.032953</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012885</td>\n",
       "      <td>0.025771</td>\n",
       "      <td>0.045594</td>\n",
       "      <td>0.065418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_5</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.105400</td>\n",
       "      <td>0.062761</td>\n",
       "      <td>0.061021</td>\n",
       "      <td>0.083210</td>\n",
       "      <td>0.105400</td>\n",
       "      <td>0.127589</td>\n",
       "      <td>0.149779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACYVA010000059.1_7</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.060398</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.060398</td>\n",
       "      <td>0.060398</td>\n",
       "      <td>0.060398</td>\n",
       "      <td>0.060398</td>\n",
       "      <td>0.060398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      count      mean       std       min       25%       50%  \\\n",
       "gene                                                                            \n",
       "CACYVA010000059.1_1     5.0  0.117536  0.148610  0.000000  0.000000  0.027985   \n",
       "CACYVA010000059.1_12   19.0  0.083405  0.047692  0.031691  0.060287  0.072677   \n",
       "CACYVA010000059.1_13    3.0  0.019184  0.016632  0.000000  0.013999  0.027999   \n",
       "CACYVA010000059.1_15    3.0  0.187434  0.063689  0.135379  0.151925  0.168472   \n",
       "CACYVA010000059.1_17    9.0  0.178138  0.044825  0.099832  0.168467  0.192534   \n",
       "CACYVA010000059.1_2    13.0  1.853915  0.896812  0.477837  1.314051  1.791888   \n",
       "CACYVA010000059.1_3     3.0  0.030396  0.032953  0.000000  0.012885  0.025771   \n",
       "CACYVA010000059.1_5     2.0  0.105400  0.062761  0.061021  0.083210  0.105400   \n",
       "CACYVA010000059.1_7     2.0  0.060398  0.000000  0.060398  0.060398  0.060398   \n",
       "\n",
       "                           75%       max  \n",
       "gene                                      \n",
       "CACYVA010000059.1_1   0.279848  0.279848  \n",
       "CACYVA010000059.1_12  0.081762  0.233605  \n",
       "CACYVA010000059.1_13  0.028777  0.029554  \n",
       "CACYVA010000059.1_15  0.213461  0.258451  \n",
       "CACYVA010000059.1_17  0.199665  0.232942  \n",
       "CACYVA010000059.1_2   2.150266  3.942154  \n",
       "CACYVA010000059.1_3   0.045594  0.065418  \n",
       "CACYVA010000059.1_5   0.127589  0.149779  \n",
       "CACYVA010000059.1_7   0.060398  0.060398  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prophage_co_exist_microdiversity.query('genome == \"GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic\"').query('prophage_gene == \"Yes\"').query('SNV_count >= 10').groupby(\"gene\").pNpS_variants.describe() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "863fe1be",
   "metadata": {},
   "outputs": [],
   "source": [
    "prophage_co_exist_microdiversity_for_plot = prophage_co_exist_microdiversity.query('genome == \"GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic\"').query('prophage_gene == \"Yes\"').query('SNV_count >= 10')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd1b292b",
   "metadata": {},
   "source": [
    "*Save MAG GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic.fasta prophage pn/ps to plot*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8f9b9bd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "prophage_co_exist_microdiversity_for_plot[['scaffold', 'gene', 'start', 'end', 'direction', 'pNpS_variants']].fillna(0).to_csv(\"/fs/ess/PAS0439/MING/virome_ecology/results/R_project/co_existing_prophage_ex_pnps.csv\", index = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "158e865c",
   "metadata": {},
   "source": [
    "*MAG GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic.fasta prophage for gggenome*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "690dbacf",
   "metadata": {},
   "outputs": [],
   "source": [
    "anno = pd.read_csv(\"../dram_anno_genomes/GCA_902777695.1_Rumen_uncultured_genome_RUG12296_genomic.fasta/annotations.tsv\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d5673e75",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_prophage_anno = anno.query('scaffold == \"CACYVA010000059.1\"').query(' 1<= gene_position <= 17')[[\"fasta\", \"scaffold\", \"gene_position\", \"start_position\", \"end_position\", \"strandedness\", \"vogdb_hit\"]]\n",
    "template_prophage_anno.to_csv(\"/fs/ess/PAS0439/MING/virome_ecology/results/R_project/co_existing_prophage_ex_for_gggenome.csv\", index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "584025d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fasta</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strandedness</th>\n",
       "      <th>vogdb_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2293</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>565</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2294</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>2</td>\n",
       "      <td>382</td>\n",
       "      <td>1026</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2295</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>3</td>\n",
       "      <td>1113</td>\n",
       "      <td>2363</td>\n",
       "      <td>1</td>\n",
       "      <td>REFSEQ reverse transcriptase; Xu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2296</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>4</td>\n",
       "      <td>2350</td>\n",
       "      <td>2769</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2297</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>5</td>\n",
       "      <td>2781</td>\n",
       "      <td>3176</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2298</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>6</td>\n",
       "      <td>3290</td>\n",
       "      <td>3550</td>\n",
       "      <td>1</td>\n",
       "      <td>REFSEQ hypothetical protein; Xu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2299</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>7</td>\n",
       "      <td>3537</td>\n",
       "      <td>3989</td>\n",
       "      <td>1</td>\n",
       "      <td>sp|P00806|ENLYS_BPT7 Endolysin; XhXp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2300</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>8</td>\n",
       "      <td>3976</td>\n",
       "      <td>4551</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2301</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>9</td>\n",
       "      <td>4553</td>\n",
       "      <td>5062</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2302</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>10</td>\n",
       "      <td>5034</td>\n",
       "      <td>5561</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2303</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>11</td>\n",
       "      <td>5660</td>\n",
       "      <td>5872</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2304</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>12</td>\n",
       "      <td>5936</td>\n",
       "      <td>10243</td>\n",
       "      <td>1</td>\n",
       "      <td>REFSEQ tail tape measure protein; Xu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2305</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>13</td>\n",
       "      <td>10246</td>\n",
       "      <td>12201</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>14</td>\n",
       "      <td>12240</td>\n",
       "      <td>13403</td>\n",
       "      <td>-1</td>\n",
       "      <td>sp|P22877|INTR_SACER Integrase; XhXr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2307</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>15</td>\n",
       "      <td>13946</td>\n",
       "      <td>14422</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2308</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>16</td>\n",
       "      <td>14422</td>\n",
       "      <td>14640</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2309</th>\n",
       "      <td>GCA_902777695.1_Rumen_uncultured_genome_RUG122...</td>\n",
       "      <td>CACYVA010000059.1</td>\n",
       "      <td>17</td>\n",
       "      <td>14752</td>\n",
       "      <td>16734</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  fasta           scaffold  \\\n",
       "2293  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2294  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2295  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2296  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2297  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2298  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2299  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2300  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2301  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2302  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2303  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2304  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2305  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2306  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2307  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2308  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "2309  GCA_902777695.1_Rumen_uncultured_genome_RUG122...  CACYVA010000059.1   \n",
       "\n",
       "      gene_position  start_position  end_position  strandedness  \\\n",
       "2293              1               2           565             1   \n",
       "2294              2             382          1026            -1   \n",
       "2295              3            1113          2363             1   \n",
       "2296              4            2350          2769             1   \n",
       "2297              5            2781          3176             1   \n",
       "2298              6            3290          3550             1   \n",
       "2299              7            3537          3989             1   \n",
       "2300              8            3976          4551             1   \n",
       "2301              9            4553          5062             1   \n",
       "2302             10            5034          5561             1   \n",
       "2303             11            5660          5872             1   \n",
       "2304             12            5936         10243             1   \n",
       "2305             13           10246         12201             1   \n",
       "2306             14           12240         13403            -1   \n",
       "2307             15           13946         14422             1   \n",
       "2308             16           14422         14640             1   \n",
       "2309             17           14752         16734            -1   \n",
       "\n",
       "                                 vogdb_hit  \n",
       "2293                                   NaN  \n",
       "2294                                   NaN  \n",
       "2295      REFSEQ reverse transcriptase; Xu  \n",
       "2296                                   NaN  \n",
       "2297                                   NaN  \n",
       "2298       REFSEQ hypothetical protein; Xu  \n",
       "2299  sp|P00806|ENLYS_BPT7 Endolysin; XhXp  \n",
       "2300                                   NaN  \n",
       "2301                                   NaN  \n",
       "2302                                   NaN  \n",
       "2303                                   NaN  \n",
       "2304  REFSEQ tail tape measure protein; Xu  \n",
       "2305                                   NaN  \n",
       "2306  sp|P22877|INTR_SACER Integrase; XhXr  \n",
       "2307                                   NaN  \n",
       "2308                                   NaN  \n",
       "2309                                   NaN  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_prophage_anno"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f270619",
   "metadata": {},
   "source": [
    "*How many samples have species which both prophage-carrying and non-carrying population co-exist?  \n",
    "For each sample, how many co-existing species?*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d80718cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "prophage_co_exist_present = prophage_co_exist_microdiversity[[\"Species\", \"id\"]].drop_duplicates().groupby(\"id\").nunique(\"Species\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "02475f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/fs/scratch/PAS0439/Ming/virome_ecology_core_prkaryotes/results/08_inStrain_rug4941/instrain_profiles/\")\n",
    "df_list = []\n",
    "for f in glob.glob(\"*\"):\n",
    "    i = f.split(\".profile\")[0]\n",
    "    path = f\"{i}.profile/output/{i}.profile_genome_info.tsv\"\n",
    "    if os.path.exists(path):\n",
    "        df = pd.read_csv(f\"{i}.profile/output/{i}.profile_genome_info.tsv\", sep = \"\\t\").query('breadth_minCov >= 0.5')   \n",
    "    else:\n",
    "        path = f\"{i}.profile/output/{i}.profile_genome_info.tsv.gz\"\n",
    "        df = pd.read_csv(f\"{path}\", sep = \"\\t\").query('breadth_minCov >= 0.5')\n",
    "    df.loc[:, \"id\"] = i\n",
    "    df_list.append(df)\n",
    "instrain_genome = pd.concat(df_list)\n",
    "instrain_genome.genome = [f.split(\".fasta\")[0] for f in list(instrain_genome.genome)]\n",
    "\n",
    "instrain_genome_taxa = pd.read_csv(\"/fs/ess/PAS0439/MING/databases/hungate_rug_rumen_mags_taxa.csv\")\n",
    "instrain_genome_final = pd.merge(instrain_genome, instrain_genome_taxa[[\"user_genome\", \"Genera\", \"Species\"]], left_on = \"genome\", right_on = \"user_genome\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4e406a15",
   "metadata": {},
   "outputs": [],
   "source": [
    "instrain_genome_present = instrain_genome_final[[\"Species\", \"id\"]].drop_duplicates().groupby(\"id\").nunique(\"Species\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "92883952",
   "metadata": {},
   "outputs": [],
   "source": [
    "prophage_co_exist_present.rename(columns = {\"Species\": \"prophage_carrying\"}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5e405c62",
   "metadata": {},
   "outputs": [],
   "source": [
    "instrain_genome_present.rename(columns = {\"Species\": \"total\"}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "97d93ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "prophage_co_exist_present_percent = pd.merge(prophage_co_exist_present, instrain_genome_present, on = \"id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "723bb8ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in prophage_co_exist_present_percent.iterrows():\n",
    "    prophage_co_exist_present_percent.loc[index,\"percent\"] = prophage_co_exist_present_percent.loc[index,\"prophage_carrying\"]/prophage_co_exist_present_percent.loc[index,\"total\"]*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6739c1d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prophage_carrying</th>\n",
       "      <th>total</th>\n",
       "      <th>percent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR3201375</th>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>2.777778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3201382</th>\n",
       "      <td>6</td>\n",
       "      <td>32</td>\n",
       "      <td>18.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3201385</th>\n",
       "      <td>3</td>\n",
       "      <td>52</td>\n",
       "      <td>5.769231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3201386</th>\n",
       "      <td>3</td>\n",
       "      <td>71</td>\n",
       "      <td>4.225352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3201388</th>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3275133</th>\n",
       "      <td>3</td>\n",
       "      <td>95</td>\n",
       "      <td>3.157895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3275135</th>\n",
       "      <td>2</td>\n",
       "      <td>105</td>\n",
       "      <td>1.904762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3275136</th>\n",
       "      <td>3</td>\n",
       "      <td>98</td>\n",
       "      <td>3.061224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3275137</th>\n",
       "      <td>6</td>\n",
       "      <td>90</td>\n",
       "      <td>6.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR3275138</th>\n",
       "      <td>13</td>\n",
       "      <td>125</td>\n",
       "      <td>10.400000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>159 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            prophage_carrying  total    percent\n",
       "id                                             \n",
       "ERR3201375                  1     36   2.777778\n",
       "ERR3201382                  6     32  18.750000\n",
       "ERR3201385                  3     52   5.769231\n",
       "ERR3201386                  3     71   4.225352\n",
       "ERR3201388                  1     50   2.000000\n",
       "...                       ...    ...        ...\n",
       "ERR3275133                  3     95   3.157895\n",
       "ERR3275135                  2    105   1.904762\n",
       "ERR3275136                  3     98   3.061224\n",
       "ERR3275137                  6     90   6.666667\n",
       "ERR3275138                 13    125  10.400000\n",
       "\n",
       "[159 rows x 3 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prophage_co_exist_present_percent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9082c8d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "prophage_co_exist_present_percent.to_csv(\"/fs/ess/PAS0439/MING/virome_ecology/results/R_project/prophage_co_exist_present_percent.csv\", index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6019f2ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Out of 580 genes, 52 are under postive selection'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "co_exist_prophage_gene_under_selection = set(prophage_co_exist_microdiversity.query('prophage_gene == \"Yes\"').query(\"divergent_site_count >=  10\").query('pNpS_variants > 1').gene)\n",
    "n_co_exist_prophage_gene_under_selection = len(co_exist_prophage_gene_under_selection)\n",
    "\n",
    "co_exist_prophage_gene_qualified = set(prophage_co_exist_microdiversity.query('prophage_gene == \"Yes\"').query(\"divergent_site_count >=  10\").gene)\n",
    "n_co_exist_prophage_gene_qualified = len(co_exist_prophage_gene_qualified)\n",
    "\n",
    "f\"Out of {n_co_exist_prophage_gene_qualified} genes, {n_co_exist_prophage_gene_under_selection} are under postive selection\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "81274f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "records = SeqIO.parse(\"/fs/scratch/PAS0439/Ming/virome_ecology_core_prkaryotes/results/08_inStrain_rug4941/repre_genomes_instrain.genes.faa\", \"fasta\")\n",
    "with open(\"/fs/scratch/PAS0439/Ming/virome_ecology_core_prkaryotes/results/08_inStrain_rug4941/prophage_carrying_genome_annotate/coexist_prophage_gene_under_selection.faa\", \"w\") as outfile:\n",
    "    for record in records:\n",
    "        if record.id in co_exist_prophage_gene_under_selection:\n",
    "            SeqIO.write(record, outfile, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cb288b0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MYENV [/fs/ess/PAS0439/MING/conda/MYENV]",
   "language": "python",
   "name": "conda_myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
