---
title: "virome_ecology"
author: "Ming Yan"
date: "10/4/2022"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: flatly
    code_download: true
---

```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

# Load libraies

```{r}
library(tidyverse)
library(vegan)
library(ggpubr)
library(rstatix)
library(phyloseq)
library(ape)
library(pairwiseAdonis)
library(data.table)
library(gggenes)
library(gghighlight)
library(cowplot)
library(gridExtra) 
```

# Load data

```{r}
raw_table <- fread('/fs/ess/PAS0439/MING/virome_ecology/results/R_project/abundance_table_raw.csv')  ## use data table to improve speed
bp <- read.csv('/fs/ess/PAS0439/MING/virome_ecology/results/R_project/samples_basepair.txt', header = FALSE, sep = '\t')  ## virus identified from contigs > 5000 bp. Therefore, the viral richness data may be covariated by the number of contig > 5000
colnames(bp) <- c("samples", "bp")
samples <- unlist(bp["samples"])
richness_raw <- colSums(raw_table[, ..samples] > 0) %>% transform()
colnames(richness_raw) <- "richness"
richness_raw <- rownames_to_column(richness_raw, var = "samples")

richness_normalized <- left_join(richness_raw, bp, by="samples")
richness_normalized["richness_normalized"] = richness_normalized["richness"]/richness_normalized["bp"]*1000000000 ## normalizing the richness profile to the same sequencing effort 

metadata_animal_performance <- read_csv("/fs/ess/PAS0439/MING/virome_ecology/results/R_project/metadata_animal_performance.csv")
metadata_ecology <- read_csv("/fs/ess/PAS0439/MING/virome_ecology/results/R_project/metadata_ecology.csv")  


table_beta_performance <- fread("/fs/ess/PAS0439/MING/virome_ecology/results/R_project/for_beta_diversity_performance.csv")

metadata_columns <- c("animal","id", "trait", "treatment", "study", "liter","breed")
metadata_columns_stats <- c("animal","id", "trait", "study", "liter")

Studies <- factor(unique(metadata_animal_performance$study))
```

# Analysis 
**Objectives: comparing viral profile in terms of alpha-diversity, beta-diversity and differential abundance analysis (DAA)**

*Data: virus richness data from 975 rumen metagenomes*

*Out of 975 samples, 312 samples from 9 studies included direct comparison of animal performence*

*Metadata include information such as ruminant species, collection method (canulated, oral tube or slaughter), breed et al*


## compare animal performance among individual studies

### Statistical test, richness
```{r}
samples_animal_performance <- metadata_animal_performance$id
richness_ready_performance <- left_join(richness_normalized, metadata_animal_performance, by=c("samples" = "id")) %>% 
  filter(samples %in% samples_animal_performance) %>% arrange(study,animal, treatment) %>% 
  mutate(treatment= factor(treatment, levels = c("High", "Medium", "Low", "HLS", "LLS", "Grazing", "TMR", "Basal", "NFC", "Protein", "Healthy", "SARA")))

stat.performance <- richness_ready_performance %>%
  group_by(liter) %>%
  wilcox_test(richness_normalized ~ treatment) %>%
  add_significance("p.adj")

## For Li et al. comparison only performed within beef and dairy

stat.performance %>% filter(!(group1 == "HLS" & group2 == "Grazing")) %>% 
  filter(!(group1 == "LLS" & group2 == "Grazing"))  %>% 
  filter(!(group1 == "HLS" & group2 == "TMR"))  %>% 
  filter(!(group1 == "LLS" & group2 == "TMR"))  
```

### Comparing richness among different animal performance traits. 
```{r}
diet_study <- c("PRJEB21624", "PRJNA492173", "PRJEB23561")

species_color <- c("#e03f34", "#24b1ed", "#e13bed", "#24ed67")
names(species_color) <- c("Beef", "Dairy", "Sheep", "Goat")

p <- richness_ready_performance %>% ggplot() + 
  aes(x=treatment, y= richness_normalized) + geom_boxplot(aes(color = animal))+
    theme(legend.position = "bottom") + labs(color = "Species") + 
  scale_color_manual(name = "Species", values = species_color) 


richness_legend <- get_legend(p)

richness_ready_performance[richness_ready_performance$study == "PRJEB23561","trait"] <- "Dietary composition (cattle)"
richness_ready_performance[richness_ready_performance$study == "PRJNA492173","trait"] <- "Dietary composition (goat)"

 richness_box_diet <- richness_ready_performance %>% filter(study %in% diet_study) %>% ggplot() + 
  aes(x=treatment, y= richness_normalized) + geom_boxplot(aes(color = animal)) +
  facet_wrap(vars(trait), scales = "free", labeller = function (labels) {   # reduce space between the facet legend 
               labels <- lapply(labels, as.character)           
               list(do.call(paste, c(labels, list(sep = "\n"))))
             }) + 
  labs(x = "", y = "Richness per billion bp") +  labs(color = "Species") +
  geom_jitter(size = 0.8, width = 0.2, height = 0.2, aes(color = animal)) + 
  scale_y_continuous(expand = expansion(mult = c(0, 0.15)))  + 
  theme_minimal()+
    theme(legend.position = "none") + 
  scale_color_manual(name = "animal", values = species_color) 


richness_box_phenotype <- richness_ready_performance %>% filter(!study %in% diet_study) %>% ggplot() + 
  aes(x=treatment, y= richness_normalized) + geom_boxplot(aes(color = animal)) +
  facet_wrap(vars(trait), scales = "free", labeller = function (labels) {   # reduce space between the facet legend 
               labels <- lapply(labels, as.character)           
               list(do.call(paste, c(labels, list(sep = "\n"))))
             }) + 
  labs(x = "", y = "Richness per billion bp") +  labs(color = "Species") +
  geom_jitter(size = 0.8, width = 0.2, height = 0.2, aes(color = animal)) + 
  scale_y_continuous(expand = expansion(mult = c(0, 0.15)))  + 
  theme_minimal()+
    theme(legend.position = "none") + 
  scale_color_manual(name = "animal", values = species_color)


## add annotations when significant
virome_richness_diet <- richness_box_diet +
   geom_curve(data=tibble(trait = "Concentrate level"), aes(x=1, xend=2, y=175, yend=175), curvature = -0.1) +
  geom_text(data=tibble(trait = "Concentrate level"), aes(x=1.5, y=190, label = "**" ), size=4) + 
  geom_curve(data=tibble(trait = "Dietary composition (goat)"), aes(x=1, xend=3, y=125, yend=125), curvature = -0.1) +
  geom_text(data=tibble(trait = "Dietary composition (goat)"), aes(x=2, y=135, label = "***" ), size=4) + 
  geom_curve(data=tibble(trait = "Dietary composition (goat)"), aes(x=2, xend=3, y=105, yend=105), curvature = -0.1) +
  geom_text(data=tibble(trait = "Dietary composition (goat)"), aes(x=2.5, y=110, label = "*" ), size=4) +
  geom_curve(data=tibble(trait = "Dietary composition (cattle)"), aes(x=1, xend=2, y=135, yend=135), curvature = -0.1) +
  geom_text(data=tibble(trait = "Dietary composition (cattle)"), aes(x=1.5, y=145, label = "***" ), size=4) +
  geom_curve(data=tibble(trait = "Dietary composition (cattle)"), aes(x=3, xend=4, y=45, yend=45), curvature = -0.1) +
  geom_text(data=tibble(trait = "Dietary composition (cattle)"), aes(x=3.5, y=55, label = "***" ), size=4) + 
  theme(text = element_text(size=8)) 
          
virome_richness_phenotype <- richness_box_phenotype +
  geom_curve(data=tibble(trait = "Methane emission"), aes(x=1, xend=2, y=102, yend=102), curvature = -0.1) +
  geom_text(data=tibble(trait = "Methane emission"), aes(x=1.5, y=107, label = "***" ), size=4) +
  geom_curve(data=tibble(trait = "Average daily gain"), aes(x=1, xend=2, y=145, yend=145), curvature = -0.1) +
  geom_text(data=tibble(trait = "Average daily gain"), aes(x=1.5, y=155, label = "*" ), size=4)   + 
  theme(text = element_text(size=8)) 

ggarrange(richness_legend, nrow = 1, heights = 0.2) # save figure legend as obj
ggsave("/fs/ess/PAS0439/MING/virome_ecology/results/R_project/virome_legend.svg", width = 5, height = 0.8)

virome_richness_diet
ggsave("/fs/ess/PAS0439/MING/virome_ecology/results/R_project/virome_richness_diet.svg", width = 4.5, height = 1.5)

virome_richness_phenotype
ggsave("/fs/ess/PAS0439/MING/virome_ecology/results/R_project/virome_richness_phenotype.svg", width = 4.5, height = 2.5)

```

### beta-diversity per study based
```{r}
# initiate empty dataframes
PCOA_df <- data.frame(matrix(ncol = 9, nrow = 0))
colnames(PCOA_df) <- c("Axis.1", "Axis.2", metadata_columns)
adonis_df <- data.frame(matrix(ncol = 13, nrow = 0))
colnames(adonis_df) <- c("pairs", "Df", "SumsOfSqs", "F.Model", "R2", "p.value", "p.adjusted", "sig", metadata_columns_stats)

permanova_df <- data.frame(matrix(ncol = 7, nrow = 0))
colnames(permanova_df) <- c("Pr(>F)", metadata_columns_stats)

variance_explained <- data.frame(matrix(ncol = 3, nrow = 0))
colnames(variance_explained) <- c("liter", "x", "y")

# Define function to perform beta diveristy analysis and permanove test.

beta_diversity <- function(Study){
  print(Study)
  dt <- table_beta_performance %>% filter(study == Study)
  bray <- vegdist(dt[ , -..metadata_columns], method = "bray")
  PCOA_metadata <- dt[ , ..metadata_columns]
  PCOA <- pcoa(bray, correction = "cailliez")
  PCOA_tmp <- as.data.frame(PCOA$vectors[, c(1,2)])
  PCOA_tmp[metadata_columns] <- PCOA_metadata  
  PCOA_df <<- rbind(PCOA_df, PCOA_tmp)  ## add result to dataframe, to plot it using faceting  
  
  variance_explained_tmp <- data.frame(matrix(ncol = 3, nrow = 1))
  colnames(variance_explained_tmp) <- c("liter", "x", "y")
  variance_explained_tmp$liter <- dt$liter[1]
  variance_explained_tmp$x <-  PCOA$values$Relative_eig[1]
  variance_explained_tmp$y <-  PCOA$values$Relative_eig[2]
  variance_explained <<- rbind(variance_explained, variance_explained_tmp)  # <<- save as global variable
  

  permanova_tmp <- adonis2(dt[ , -..metadata_columns] ~ dt$treatment)$`Pr(>F)`[1] %>% as.data.frame()
  colnames(permanova_tmp) <- "Pr(>F)"
  adonis_tmp <- pairwise.adonis(dt[ , -..metadata_columns], dt$treatment) %>% as.data.frame() ## pairwised adonis
  
  adonis_metadata <- dt[1, ..metadata_columns_stats]
  adonis_tmp[metadata_columns_stats] <- adonis_metadata
  adonis_df <<- rbind(adonis_df, adonis_tmp)
  
  permanova_tmp[metadata_columns_stats] <- adonis_metadata  ## add result to dataframe, to plot it using faceting
  permanova_df <<- rbind(permanova_df, permanova_tmp) 
  
}


# for Li et al., 2019 data was collected from three different breeds, considered as blocks
beta_diversity_with_block <- function(Study){
  print(Study)
  dt <- table_beta_performance %>% filter(study == Study)
  bray <- vegdist(dt[ , -..metadata_columns], method = "bray")
  PCOA_metadata <- dt[ , ..metadata_columns]
  PCOA <- pcoa(bray, correction = "cailliez")
  PCOA_tmp <- as.data.frame(PCOA$vectors[, c(1,2)])
  PCOA_tmp[metadata_columns] <- PCOA_metadata  
  PCOA_df <<- rbind(PCOA_df, PCOA_tmp)  ## add result to dataframe, to plot it using faceting  
  
  variance_explained_tmp <- data.frame(matrix(ncol = 3, nrow = 1))
  colnames(variance_explained_tmp) <- c("liter", "x", "y")
  variance_explained_tmp$liter <- dt$liter[1]
  variance_explained_tmp$x <-  PCOA$values$Relative_eig[1]
  variance_explained_tmp$y <-  PCOA$values$Relative_eig[2]
  variance_explained <<- rbind(variance_explained, variance_explained_tmp)
  
  permanova_tmp <- adonis2(dt[ , -..metadata_columns] ~ dt$treatment, strata = dt$breed)$`Pr(>F)`[1] %>% as.data.frame()
  colnames(permanova_tmp) <- "Pr(>F)"
  adonis_tmp <- pairwise.adonis(dt[ , -..metadata_columns], dt$treatment) %>% as.data.frame() ## pairwised adonis
  
  adonis_metadata <- dt[1, ..metadata_columns_stats]
  adonis_tmp[metadata_columns_stats] <- adonis_metadata
  adonis_df <<- rbind(adonis_df, adonis_tmp)
  
  permanova_tmp[metadata_columns_stats] <- adonis_metadata  ## add result to dataframe, to plot it using faceting
  permanova_df <<- rbind(permanova_df, permanova_tmp) 
  
}


for (Study in Studies){
  if (Study == "PRJNA448333"){   ## different breeds used, as blocks
    beta_diversity_with_block(Study)
  } else{
  beta_diversity(Study)
}}

```

### adonis
```{r}
permanova_df %>% select(c("Pr(>F)", animal, trait, liter)) %>% print() %>% as_tibble()

```

### pairwised adonis
```{r}
adonis_df %>% select(c(pairs, p.adjusted, animal, trait, liter)) %>% print() %>% as.tibble()
```

### plot pcoa
### extract lagend and plot it individually
```{r}
PCOA_df[PCOA_df$study == "PRJEB23561","trait"] <- "Dietary composition (cattle)"
PCOA_df[PCOA_df$study == "PRJNA492173","trait"] <- "Dietary composition (goat)"
PCOA_df[PCOA_df$study == "PRJNA597489","trait"] <- "Feed efficiency (milk)"
PCOA_df[PCOA_df$study == "PRJNA448333","trait"] <- "Feed efficiency (meat)"

PCOA_df <- left_join(PCOA_df %>% select(-treatment), metadata_animal_performance[, c("id", "treatment")], by = "id")

function_pcoa <- function(literature){ PCOA_df %>% filter(liter == literature) %>% ggplot(aes(x=Axis.1, y=Axis.2, color=treatment)) + geom_point() + facet_wrap(vars(trait), labeller = function (labels) {   # reduce space between the facet legend 
               labels <- lapply(labels, as.character)           # https://stackoverflow.com/questions/41311810/how-to-reduce-vertical-spacing-between-facet-labels-when-using-facet-wrap
               list(do.call(paste, c(labels, list(sep = "\n"))))
             }) +
  theme(legend.position="none", axis.ticks.y=element_blank(), 
        axis.ticks.x=element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_blank())  +
  scale_x_continuous(name = paste("Axis 1 (", round(variance_explained[variance_explained$liter==literature,"x"]*100, 2), "%)", sep = "")) +
  scale_y_continuous(name = paste("Axis 2 (", round(variance_explained[variance_explained$liter==literature,"y"]*100, 2), "%)", sep = "")) +
   theme_minimal() +
    theme(legend.position = "bottom") +
    theme(legend.title = element_blank(), 
               text = element_text(size=8)) +
    theme(legend.margin=margin(0,0,0,0),  ## reduce space between the legend and the figure
        legend.box.margin=margin(-10,-10,-10,-10))
}

pcoa_plots <- lapply(unique(PCOA_df$liter), function_pcoa)

names(pcoa_plots) <- unique(PCOA_df$liter)

pcoa_plots[["Shi et al., 2014"]] <- pcoa_plots[["Shi et al., 2014"]] + annotate("text", -Inf, -Inf, hjust = 0, vjust = 0, label = "Permanova: ***", size = 2)  # hjust and vjust are defined between 0 and 1, min to max
pcoa_plots[["Li et al., 2019"]] <- pcoa_plots[["Li et al., 2019"]] + annotate("text", -Inf, -Inf, hjust = 0, vjust = 0, label = "Permanova: ***", size = 2)
pcoa_plots[["Li et al., 2020"]] <- pcoa_plots[["Li et al., 2020"]] + annotate("text", -Inf, -Inf, hjust = 0, vjust = 0, label = "Permanova: ***", size = 2) 
pcoa_plots[["Stewart et al., 2018"]] <- pcoa_plots[["Stewart et al., 2018"]] + annotate("text",  -Inf, -Inf, hjust = 0, vjust = 0, label = "Permanova: ***", size = 2) 
pcoa_plots[["Xue et al., 2020"]] <- pcoa_plots[["Xue et al., 2020"]] + annotate("text",  -Inf, -Inf, hjust = 0, vjust = 0, label = "Permanova: *", size = 2) 
pcoa_plots[["Lu et al., 2019"]] <- pcoa_plots[["Lu et al., 2019"]] + annotate("text",  -Inf, -Inf, hjust = 0, vjust = 0, label = "Permanova: ***", size = 2) 
 
#pcoa_final <- plot_grid(pcoa_plots[[1]], pcoa_plots[[2]],pcoa_plots[[3]],pcoa_plots[[4]],pcoa_plots[[5]],pcoa_plots[[6]],pcoa_plots[[7]],pcoa_plots[[8]],pcoa_plots[[9]])

pcoa_diet <- ggarrange(pcoa_plots$`Stewart et al., 2018`, pcoa_plots$`Li et al., 2020`, pcoa_plots$`Lu et al., 2019`, nrow = 1) #+ labs(tag = "A") + 
  #theme(plot.tag.position = c(0, 0.97))

pcoa_phenotype <- ggarrange(pcoa_plots$`Anderson et al., 2021`, pcoa_plots$`Shi et al., 2014`, pcoa_plots$`Xue et al., 2020`, pcoa_plots$`Stergiadis et al., 2021`, pcoa_plots$`Li et al., 2019`, pcoa_plots$`Xue et al., 2022`) #+ labs(tag = "B") + 
 # theme(plot.tag.position = c(0, 0.97))

## adjust tag position, https://stackoverflow.com/questions/54207695/put-tag-without-losing-space-for-actual-chart 

pcoa_final <- ggarrange(pcoa_diet, pcoa_phenotype, nrow = 2, heights = c(1, 2)) 
pcoa_final
ggsave("/fs/ess/PAS0439/MING/virome_ecology/results/R_project/virome_pcoa.svg", width = 7, height = 5)
```


# session info
```{r}
sessionInfo()
```
